softlearning/environments/adapters/dm_control_adapter.py

```diff
@@ -0,0 +1,144 @@
+        return gym_box
+    elif isinstance(dm_control_space, ArraySpec):
+        return spaces.Box(
+            low=-float("inf"),
+            high=float("inf"),
+            shape=dm_control_space.shape,
+            dtype=dm_control_space.dtype)
+    elif isinstance(dm_control_space, OrderedDict):
+        return spaces.Dict(OrderedDict([
+            (key, convert_dm_control_to_gym_space(value))
+            for key, value in dm_control_space.items()
+        ]))
+    else:
+        raise ValueError(dm_control_space)
+
+
+class DmControlAdapter(SoftlearningEnv):
+    """Adapter that implements the SoftlearningEnv for DeepMind Control Suite envs."""
+
+    def __init__(self,
+                 domain,
+                 task,
+                 *args,
+                 env=None,
+                 normalize=True,
+                 observation_keys=None,
+                 unwrap_time_limit=True,
+                 **kwargs):
+        assert not args, (
+            "Gym environments don't support args. Use kwargs instead.")
+
+        self.normalize = normalize
+        self.unwrap_time_limit = unwrap_time_limit
+
+        self._Serializable__initialize(locals())
+        super(DmControlAdapter, self).__init__(domain, task, *args, **kwargs)
+        if env is None:
+            assert (domain is not None and task is not None), (domain, task)
+            env = suite.load(
+                domain_name=domain,
+                task_name=task,
+                task_kwargs=kwargs
+                # TODO(hartikainen): Figure out how to pass kwargs to this guy.
+                # Need to split into `task_kwargs`, `environment_kwargs`, and `visualize_reward` bool.
+                # Check the suite.load(.) in:
+                # https://github.com/deepmind/dm_control/blob/master/dm_control/suite/__init__.py
+            )
+        else:
+            assert domain is None and task is None, (domain, task)
+
+        assert isinstance(env.observation_spec(), OrderedDict)
+        self.observation_keys = (
+            observation_keys or tuple(env.observation_spec().keys()))
+        # TODO(hartikainen): Do we need to normalize actions? For now, assume
+        # we don't.
+
+        self._env = env
+
+    @property
+    def observation_space(self):
+        observation_space = convert_dm_control_to_gym_space(self._env.observation_spec())
+        return observation_space
+
+    @property
+    def active_observation_shape(self):
+        """Shape for the active observation based on observation_keys."""
+        observation_space = self.observation_space
+        active_size = sum(
+            np.prod(observation_space.spaces[key].shape)
+            for key in self.observation_keys)
+        active_shape = (int(active_size), )
+        return active_shape
+
+    def convert_to_active_observation(self, observation):
+        flattened_observation = np.concatenate([
+            observation[key] for key in self.observation_keys])
```

Comment:
Let's add explicit axis for this concatenate. Axis should be -1. https://docs.scipy.org/doc/numpy/reference/generated/numpy.concatenate.html

Reply:
Thanks. And is there a reason why we prefer `axis=-1` as opposed to `axis=1` because I think both do the same thing for 2D arrays?

Reply:
`axis=1` would work fine too, and might even be less error-prone. Having `axis=-1` just makes it more explicit that it's the last axis.